include/range/v3/view/subrange.hpp

```diff
@@ -0,0 +1,428 @@
+                return last_();
+            }
+            constexpr bool empty() const
+            {
+                return first_() == last_();
+            }
+
+            CONCEPT_REQUIRES(K == subrange_kind::sized)
+            RANGES_CXX14_CONSTEXPR difference_type_t<I> size() const
+            {
+#if RANGES_CXX_IF_CONSTEXPR >= RANGES_CXX_IF_CONSTEXPR_17
+                if constexpr (StoreSize) {
+                    return size_();
+                } else {
+                    return last_() - first_();
+                }
+#else
+                return size_helper(meta::bool_<StoreSize>());
+#endif
+            }
+
+            RANGES_NODISCARD RANGES_CXX14_CONSTEXPR
+            subrange next(difference_type_t<I> n = 1) const
+            {
+                auto tmp = *this;
+                tmp.advance(n);
+                return tmp;
+            }
+            CONCEPT_REQUIRES(BidirectionalIterator<I>())
+            RANGES_NODISCARD RANGES_CXX14_CONSTEXPR
+            subrange prev(difference_type_t<I> n = 1) const
+            {
+                auto tmp = *this;
+                tmp.advance(-n);
+                return tmp;
+            }
+            RANGES_CXX14_CONSTEXPR subrange &advance(difference_type_t<I> n)
+            {
+                difference_type_t<I> remainder = ranges::advance(first_(), n, last_());
+#if RANGES_CXX_IF_CONSTEXPR >= RANGES_CXX_IF_CONSTEXPR_17
+                (void)remainder;
+                if constexpr (StoreSize)
+                    size_() -= n - remainder;
+#else
+                modify_size(remainder - n, meta::bool_<StoreSize>{});
+#endif
+                return *this;
+            }
+        };
+
+        template<typename I, typename S, subrange_kind K>
+        struct is_referenceable_range<subrange<I, S, K>> : std::true_type {};
+
+        template<typename I, typename S = I, CONCEPT_REQUIRES_(Sentinel<S, I>())>
+        using sized_subrange = subrange<I, S, subrange_kind::sized>;
+
+#if RANGES_CXX_DEDUCTION_GUIDES >= RANGES_CXX_DEDUCTION_GUIDES_17
+        template<typename I, typename S, CONCEPT_REQUIRES_(Sentinel<S, I>())>
+        subrange(I, S, difference_type_t<I>) -> subrange<I, S, subrange_kind::sized>;
+
+        template<typename P, CONCEPT_REQUIRES_(detail::IteratorSentinelPair<P>())>
+        subrange(P) -> subrange<
+            meta::_t<std::tuple_element<0, P>>,
+            meta::_t<std::tuple_element<1, P>>>;
+
+        template<typename P, CONCEPT_REQUIRES_(detail::IteratorSentinelPair<P>())>
+        subrange(P, difference_type_t<meta::_t<std::tuple_element<0, P>>>) ->
+            subrange<meta::_t<std::tuple_element<0, P>>,
+                meta::_t<std::tuple_element<1, P>>, subrange_kind::sized>;
+
+        template<typename R, CONCEPT_REQUIRES_(ReferenceableRange<R>())>
+        subrange(R &&) -> subrange<iterator_t<R>, sentinel_t<R>,
+            static_cast<subrange_kind>(SizedRange<R>() ||
+                SizedSentinel<sentinel_t<R>, iterator_t<R>>())>;
```

Comment:
This disjunction is theoretically necessary: a single-pass range could satisfy `SizedSentinel<sentinel_t<R>, iterator_t<R> && !SizedRange<R>` since `ranges::size` cannot determine the range size from `end() - begin()` for non-forward ranges. A range could legitimately overload `size` to return something other than the size of the range to meet some strange external requirement, and specialize `disable_sized_range`.

Reply:
A comment in the code to that effect would be a nice addition.